opencog/atomspace/AtomTable.h

```diff
@@ -232,12 +242,26 @@ class AtomTable
                         bool subclass = false) const
     {
         std::lock_guard<std::recursive_mutex> lck(_mtx);
+
+        // We're going to use this one handle so we don't construct/destruct
+        // googles of temporaries in the process of doing the copies.
+        Handle temp_handle;
         std::for_each(typeIndex.begin(type, subclass),
                       typeIndex.end(),
-             [&](Handle h)->void {
-                  if (not isDefined(h)) return;
-                  (func)(h);
-             });
+            [&](UUID uuid)->void {
+                if (UNDEFINED_UUID == uuid) return;
+                
+                // Since the function could and probably will resolve the
+                // AtomPtr, we need to reset the handle since we are going to
+                // bypass the normal shared_ptr semantics here to avoid
+                // a copy per handle which is much more expensive than
+                // resetting and setting a new UUID. So we call the safe
+                // version which takes a bit longer.
+                temp_handle.set_uuid(uuid);
```

Comment:
No, copying is most certainly much less than resolve.

Except, typeIndex and targetTypeIndex don't hold handles, they hold UUIDs. They return Handles that they create for the express purpose of returning them to this iterator function. It is that creation that I have eliminated, and the subsequent copying on the stack.

In this particular case of iterating over types to get atoms, we're going to need to do a resolve no matter what we do, unless we change typeIndex and targetTypeIndex to use Handles. That is a much bigger change and one I didn't want to even propose without talking to you. I don't know if that would make things faster or slower as I don't know how often these routines are actually used for long-running processes like pattern matching and such. As I learn more, I'll be able to make better proposals.